### PR TITLE
Upgrade to Jetty 12 for Jakarta build

### PR DESCRIPTION
**Pull request by dalbani**
Spring Boot 3.2, which has been released recently, uses Jetty 12.
So in order to support integrating Scalatra in Spring Boot 3.2 applications, this MR aims at upgrading Jetty.

### CODE REVIEW

**Comment by dalbani** [comment]
I take the liberty to ping you, @xuwei-k, as you worked on the `javax/jakarta` build matrix.
I originally thought the upgrade to Jetty 12 would be a drop-in replacement of Jetty 11, but they've actually changed the package name and some part of the API (in particular regarding resources).

I've managed to get "something" to compile with the introduction of an additional "compatibility class" (`JettyCompat`).
I had to make `scalatra-test` depend on `scalatra-jetty` to have access to this `JettyCompat`, instead of on `scalatra-common` previously.
But here's the catch: `scalatra-jetty` depends on `scalatra`, which in turns depends on `scalatra-test`.
In other words: that's a cyclic dependency.

In the meantime, for the code to compile, I made `scalatra-jetty` depend on `scalatra-common` but that's obviously not enough, as `ScalatraServlet` is not accessible anymore for example.

Any idea on how to solve that?

A bolder solution would be to drop the `javax` build and only keep the `jakarta` one. What do you think of that?
I think it's now time for everytime to move to Jakarta EE 🤓

**Comment by takezoe** [comment]
We know that Scalatra has some existing users who still need to stay at javax. Dropping javax can't be an option at this time.

**Comment by dalbani** [comment]
OK, good to know.

In https://github.com/scalatra/scalatra/pull/1618, I managed to get Scalatra to work with Jetty 12 and Jakarta EE 10.
Though that meant lifting the Java requirement to 17 (see https://eclipse.dev/jetty/#:~:text=Java-,Jetty%2012.0.x,-6.0%2C%205.0%2C%204.0).
Is that a problem for the existing users that you mentioned?

So, all in all, it looks the remaining issue to have `javax` (Jetty 10) and `jakarta` (Jetty 12) in the same codebase.
I think it's "only" a matter of fixing the module dependency issue I mentioned initially.
If someone wants to help me out on this, I'd greatly appreciate it.

**Comment by dalbani** [comment]
I think I've found a solution to the cyclic dependency issue I was referring to in my previous message.
I introduced a new module called `scalatra-compat` which includes the previously existing `ServletCompat` and the newly added `JettyCompat`.

All the test pass according to the CI checks.
Could you have a look, @takezoe?

I still need to re-enable MiMa tests, due to changes in the `scalatra`, `scalatra-common`, `scalatra-jetty` and `scalatra-test` modules which are causing the following problems:
> `MissingClassProblem`, `DirectMissingMethodProblem`, `ReversedMissingMethodProblem`, `IncompatibleResultTypeProblem`

Should I simply add the appropriate `ProblemFilters.exclude` instructions?

File: .github/workflows/scala.yml [hunk 1]
```diff
@@ -21,5 +21,5 @@ jobs:
         distribution: "temurin"
         java-version: ${{ matrix.java }}
     - name: Run tests
-      run: sbt -v mimaReportBinaryIssues test unidoc
+      run: sbt -v test unidoc
```
Comment (dalbani) [inline .github/workflows/scala.yml:24]: Mima tests needs to be restored (see PR discussion).

File: build.sbt [hunk 2]
```diff
@@ -58,18 +58,21 @@ lazy val scalatraSettings = Seq(
       case Some((2, _)) =>
         unusedOptions ++ Seq(
           "-Xsource:3",
-          "-release:8",
+          "-release:" + {
```
Comment (dalbani) [inline build.sbt:61]: I don't have much experience with SBT and those "virtual axes", so tell me if what I wrote doesn't make sense 😄 
What I was trying to do was to compile to Java 17 bytecode for the Jakarta build; otherwise stay with Java 8 as before.

File: build.sbt [hunk 3]
```diff
@@ -58,18 +58,21 @@ lazy val scalatraSettings = Seq(
...
+            }
+          },
           "-Xlint",
           "-Xcheckinit",
         )
       case _ =>
         Nil
     }
   },
-  javacOptions ++= Seq(
```
Comment (dalbani) [inline build.sbt:69]: I got rid of this section because, from what I understand, this section is only used to compile `.java` files, isn't it?

**Comment by dalbani** [comment]
What do you think, @xuwei-k and @takezoe, of my implementation in this MR? Thanks!

File: core/src/main/scala/org/scalatra/servlet/FileUploadSupport.scala [hunk 4]
```diff
@@ -73,6 +74,7 @@ trait FileUploadSupport extends ServletBase with HasMultipartConfig {
    * `IllegalStateException` for some other reason.
    */
   protected def isSizeConstraintException(e: Exception): Boolean = e match {
+    case exc: ServletException => exc.getRootCause.getMessage == "400: bad multipart"
```
Comment (takezoe) [inline core/src/main/scala/org/scalatra/servlet/FileUploadSupport.scala:77]: Is this related to upgrading to Jetty12? Otherwise, better to separate PR.

**Review by takezoe — COMMENTED** [review]
@dalbani Sorry for late reaction. Basically this looks good to me but added some comments. Could you take a look?

File: core/src/test/scala/org/scalatra/CookieSupportTest.scala [hunk 5]
```diff
@@ -131,15 +131,15 @@ class CookieSupportTest extends ScalatraFunSuite {
     post("/foo/remove-cookie") {
       val hdr = response.getHeader("Set-Cookie")
       // Jetty turns Max-Age into Expires
-      hdr should include("; Expires=Thu, 01-Jan-1970 00:00:00 GMT")
+      hdr should include regex "; Expires=Thu, 01(-| )Jan(-| )1970 00:00:00 GMT"
```
[the remark this answers is not in the document]
Comment (dalbani) [reply core/src/test/scala/org/scalatra/CookieSupportTest.scala:134]: Indeed, that's the reason.

File: core/src/main/scala/org/scalatra/servlet/FileUploadSupport.scala (hunk 4 above)
Comment (dalbani) [reply core/src/main/scala/org/scalatra/servlet/FileUploadSupport.scala:77]: It is indeed related to the Jetty 12.
See https://github.com/jetty/jetty.project/blob/jetty-12.0.x/jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiRequest.java#L591 for the reason of this change.

File: core/src/main/scala-jakarta-jvm/org/scalatra/servlet/HttpServletRequestReadOnly.scala [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+ *
+ * Limitations of this class include the following:
+ *
+ *   - it is mostly immutable (methods on the original request are not given stable values,
+ *     nor are methods that return non-primitive types)
+ *   - changes made to the original object or this object may not be reflected across threads
+ *
+ * @param originalRequest the original HttpServletRequest to wrap
+ */
+case class HttpServletRequestReadOnly(private val originalRequest: HttpServletRequest) extends HttpServletRequestWrapper(originalRequest) {
```
[the remark this answers is not in the document]
Comment (dalbani) [reply core/src/main/scala-jakarta-jvm/org/scalatra/servlet/HttpServletRequestReadOnly.scala:20]: Version 6 of the Jakarta Servlet API has indeed removed methods, which explains this incompatibility.

File: build.sbt (hunk 2 above)
Comment (dalbani) [reply build.sbt:61]: It has just occurred to me that I put this `-release` argument at the wrong location, under `unusedOptions` 🙄 
But looking at this question of the Java bytecode version, that made me think: shouldn't Scalatra simply be compiled for Java 8 for _all_ builds?
As the version requirement is actually a *runtime* requirement: the `javax` build requires 8+, while the `jakarta` build requires 17+ due to Jetty 12.
What do you think?

File: .github/workflows/scala.yml [hunk 7]
```diff
@@ -8,7 +8,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        java: [11, 21]
+        java: [17, 21]
```
Comment (dalbani) [inline .github/workflows/scala.yml:11]: As a follow-up to my remark in https://github.com/scalatra/scalatra/pull/1610/files#r1438577326, would you know a way to to run `sbt test` for a specific (i.e. `javax` or `jakarta`) build only?
So that the `javax` build would tested against Java 8, 11, 17 and 21, while the `jakarta` build would be tested against Java 17 and 21 only.

Comment (dalbani) [reply .github/workflows/scala.yml:11]: I had an idea: what about enabling the `jakarta` axis in `build.sbt` only in case SBT is running in a JDK 17+ environment?

File: core/src/main/scala/org/scalatra/servlet/FileUploadSupport.scala (hunk 4 above)
Comment (takezoe) [reply core/src/main/scala/org/scalatra/servlet/FileUploadSupport.scala:77]: Got it. Thanks for clarification.

File: core/src/test/scala/org/scalatra/CookieSupportTest.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (takezoe) [reply core/src/test/scala/org/scalatra/CookieSupportTest.scala:134]: Thank. Looks like whitespace is a more preferable format in HTTP spec.
https://datatracker.ietf.org/doc/html/rfc7231#section-7.1.1.1

File: build.sbt (hunk 2 above)
Comment (takezoe) [reply build.sbt]: If Java 17+ for Jetty 12 is the only runtime requirement as you mentioned, targeting Java 11 is better because backward compatibility is one of Scalatra's highest priority principals.

File: .github/workflows/scala.yml (hunk 7 above)
Comment (takezoe) [reply .github/workflows/scala.yml:11]: As I commented above, I think we can target Java 17+ in Scalatra 3.1.x if we would upgrade Jetty to 12, and keeping 3.0.x maintained for users who still need Java 11.

**Comment by takezoe** [comment]
@dalbani I updated my comments again. In short, I think we can target Java 17+ in Scalatra 3.1.0.

File: .github/workflows/scala.yml (hunk 7 above)
Comment (dalbani) [reply .github/workflows/scala.yml]: Oh, I thought you didn't want to maintain the 3.0.x branch.
But I think I have a solution to my original question:
> So that the `javax` build would tested against Java 11, 17 and 21, while the `jakarta` build would be tested against Java 17 and 21 only.

I have implemented a "trick" in [Run tests for "javax" build on JDK 11 (only)](https://github.com/scalatra/scalatra/pull/1610/commits/cb9d2adc0184d7fffc0a771e24650d729131cc77) in order to implement that.
What do you think?

File: core/src/main/scala/org/scalatra/servlet/FileUploadSupport.scala (hunk 4 above)
Comment (dalbani) [reply core/src/main/scala/org/scalatra/servlet/FileUploadSupport.scala]: I have added some comments regarding specificities related to Jetty 10 and 12.
(I've also refined the way how those exceptions are "recognized".)

File: build.sbt (hunk 2 above)
Comment (dalbani) [reply build.sbt]: See my remark in https://github.com/scalatra/scalatra/pull/1610#discussion_r1438942324, which proposes a solution where Scalatra 3.1.x could keep compatibility with Java 11 (`javax` build only, obviously).

File: build.sbt [hunk 8]
```diff
@@ -78,7 +81,8 @@ lazy val scalatraSettings = Seq(
     "-feature",
     "-language:higherKinds",
     "-language:implicitConversions",
-    "-language:existentials"
+    "-language:existentials",
+    "-release:" + javaTargetRelease(virtualAxes.?.value.getOrElse(Nil)),
```
Comment (dalbani) [inline build.sbt:85]: This change (and the accompanying one above for `javacOptions`) sets the appropriate compilation target, depending on whether this is a `javax` build (11) or a `jakarta` build (17).

Note: I just learned that Scala 2.12 is only able to emit Java 8 bytecode.
So, in this case, `-release` means "compile against a specific platform API version".
So my question is: what about dropping support for Scala 2.12?

File: project/SbtWhen.scala [hunk 9]
```diff
@@ -0,0 +1,7 @@
+object SbtWhen {
+  implicit class When[A](a: A) {
```
Comment (dalbani) [inline project/SbtWhen.scala:2]: Shamelessly ~~stolen~~ inspired from https://stackoverflow.com/questions/7313604/conditional-invocation-of-a-method-in-scala...
I'm no Scala nor SBT expert, but any improvement is welcome.

File: build.sbt (hunk 8 above)
Comment (dalbani) [reply build.sbt:85]: Regarding the compilation target to 17 for the `jakarta`, I leave it to your appreciation.
While Jetty 12 requires Java 17, I just realized that, for example, Tomcat 10.1.x does support Jakarta Servlet 6, yet with Java 11 only as requirement (https://tomcat.apache.org/whichversion.html).
So let me know if you prefer to stick to Java 11 for compatibility reasons.

File: core/src/main/scala/org/scalatra/servlet/FileUploadSupport.scala (hunk 4 above)
Comment (takezoe) [reply core/src/main/scala/org/scalatra/servlet/FileUploadSupport.scala]: Thank you!

File: .github/workflows/scala.yml (hunk 7 above)
Comment (takezoe) [reply .github/workflows/scala.yml]: > Oh, I thought you didn't want to maintain the 3.0.x branch.

Haha. Yeah, I don't want to maintain multiple branches if possible, but we always had to do it because when we upgrade Scalatra version these days, there were always incompatible changes like dropping Scala version or Java version. We needed to maintain old version for users who need it.

This time, I only think that dropping `javax` is too early.

File: build.sbt (hunk 8 above)
Comment (takezoe) [reply build.sbt:85]: Oh, I see. It may be time to drop Scala 2.12 support.

Comment (takezoe) [reply build.sbt:85]: It would be great if Scalatra 3.1.0 can support Java 11 both for `javax` and `jakarta`.

File: .github/workflows/scala.yml (hunk 7 above)
Comment (dalbani) [reply .github/workflows/scala.yml]: Up to you to decide 😄 
I shall revert/leave the code as you wish.

File: build.sbt (hunk 8 above)
Comment (dalbani) [reply build.sbt:85]: > It would be great if Scalatra 3.1.0 can support Java 11 both for javax and jakarta.

In this case, I would only need to revert https://github.com/scalatra/scalatra/pull/1610/commits/a475aae4d8e8791e14da6c4c8f6526de4decd5aa and Java 11 would be supported for both `javax` and `jakarta` (albeit the last one with a container different than Jetty 12; Tomcat 10.1.x for example, as I mentioned above).

Comment (dalbani) [reply build.sbt]: As I alluded to in my previous message, I reverted https://github.com/scalatra/scalatra/commit/a475aae4d8e8791e14da6c4c8f6526de4decd5aa, in order to support a Jakarta deployment with Tomcat 10.1.x on Java 11.

Speaking of Tomcat, I've just added a new module to Scalatra, which will give the ability to *test* the Jakarta build on Java 11.
What do you think of that?

File: .github/workflows/scala.yml (hunk 7 above)
Comment (dalbani) [reply .github/workflows/scala.yml]: For the record, I've restricted the use of the `runningOnJvm17AndAbove()` check to the `jetty` module when used in the `jakarta` build.

Comment (dalbani) [reply .github/workflows/scala.yml]: I stand correct on my last comment. It's also the `test` module that has to be excluded when running on Java 11, because this module internally relies on Jetty via `[Embedded]JettContainer`.

I suppose we could introduce a `[Embedded]TomcatContainer` and use it when the test suite is run on Java 11?
But I suppose that would require some "magic" (macro?) to dynamically choose between Jetty and Tomcat.
Is that an option that is worth exploring?

File: build.sbt (hunk 8 above)
Comment (takezoe) [reply build.sbt]: Okay, our options are:

- Drop Java 11 support
- Continue Java 11 support with Tomcat module

However, still some sort of workaround seems to be necessary to run CI with Java 11 for Scalatra itself and it will anyway bring complexity I"m concerning.

I think it's okay to drop Java 11 support in Scalatra 3.1.0. Upgrading Java version would be relatively easier than migration applications from javax to jakarta.

Comment (dalbani) [reply build.sbt]: > I think it's okay to drop Java 11 support in Scalatra 3.1.0. Upgrading Java version would be relatively easier than migration applications from javax to jakarta.

So, just for confirmation, dropping Java 11 support in the sense of compiling for Java 17?
Or keep compiling for Java 11 but stop actively testing (in CI) if that Scalatra still works in that environment?

The second option would be like:
> Scalatra 3.1.0 officially supports Java 17, but _may_ work with Java 11 (i.e. with an application server that accepts it, as long as it supports Java Servlet 4.0 or Jakarta Servlet 6, depending on the chosen build).

Comment (dalbani) [reply build.sbt]: Hey @takezoe, did you see my last message in this thread?
Please indicate which of the 2 option I should go forward with.

Comment (dalbani) [reply build.sbt]: Hey @takezoe, I hope everything's going fine on your side.
Did you have the opportunity to look at my last message?

File: .github/workflows/scala.yml (hunk 7 above)
Comment (takezoe) [reply .github/workflows/scala.yml]: Hmm, right. Or simply using `EmbeddedTomcatContainer` in tests could be an option. But again, I don't mind about dropping Java 11 support in Scalatra 3.1.

File: build.sbt (hunk 8 above)
Comment (takezoe) [reply build.sbt]: Sorry for my late response as my new year vacation awas ended and got back to work.

> So, just for confirmation, dropping Java 11 support in the sense of compiling for Java 17?

This is what I meant.

File: project/SbtWhen.scala (hunk 9 above)
Comment (takezoe) [reply project/SbtWhen.scala:2]: Haha, I'm also very good at SBT. Looks good if we actually need switching by condition for now.

File: build.sbt (hunk 8 above)
Comment (dalbani) [reply build.sbt]: OK, following your input, I've created a new PR with just the right commits: https://github.com/scalatra/scalatra/pull/1632.
If you agree, we can close this one and use the new one instead.

Comment (takezoe) [reply build.sbt]: Thank you! https://github.com/scalatra/scalatra/pull/1632 looks good. I just added one comment about MiMa to the new PR.

**Comment by dalbani** [comment]
Closing in favour of https://github.com/scalatra/scalatra/pull/1632.
